core/src/main/java/com/linecorp/armeria/internal/common/BodySubscriber.java

```diff
@@ -0,0 +1,306 @@
+        }
+    }
+
+    @Override
+    public void onNext(HttpObject httpObject) {
+        if (httpObject instanceof HttpHeaders) {
+            final HttpHeaders trailers = (HttpHeaders) httpObject;
+            completeTrailers(trailers);
+            return;
+        }
+
+        final Subscriber<? super HttpData> downstream = this.downstream;
+        assert downstream != null;
+        assert httpObject instanceof HttpData;
+
+        final EventExecutor executor = this.executor;
+        if (executor.inEventLoop()) {
+            onNext0((HttpData) httpObject);
+        } else {
+            executor.execute(() -> onNext0((HttpData) httpObject));
+        }
+    }
+
+    private void onNext0(HttpData httpData) {
+        wroteAny = true;
+        if (!usePooledObject) {
+            httpData = PooledObjects.copyAndClose(httpData);
+        }
+        downstream.onNext(httpData);
+    }
+
+    @Override
+    public void onComplete() {
+        final EventExecutor executor = this.executor;
+        final Subscriber<? super HttpData> downstream = this.downstream;
+        if (executor == null || downstream == null) {
+            completing = true;
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            onComplete0(downstream);
+        } else {
+            executor.execute(() -> onComplete0(downstream));
+        }
+    }
+
+    private void onComplete0(Subscriber<? super HttpData> downstream) {
+        downstream.onComplete();
+    }
+
+    @Override
+    public void onError(Throwable cause) {
+        final EventExecutor executor = this.executor;
+        final Subscriber<? super HttpData> downstream = this.downstream;
+        if (executor == null || downstream == null) {
+            this.cause = cause;
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            onError0(cause, downstream);
+        } else {
+            executor.execute(() -> onError0(cause, downstream));
+        }
+    }
+
+    private void onError0(Throwable cause, Subscriber<? super HttpData> downstream) {
+        downstream.onError(cause);
+        this.downstream = NoopSubscriber.get();
+    }
+
+    /**
+     * Completes the specified trailers.
+     */
+    protected final void completeTrailers(HttpHeaders trailers) {
+        HeadersFuture<HttpHeaders> trailersFuture = this.trailersFuture;
+        if (trailersFuture != null) {
+            trailersFuture.doComplete(trailers);
+            return;
+        }
+
+        trailersFuture = new HeadersFuture<>();
+        if (trailersFutureUpdater.compareAndSet(this, null, trailersFuture)) {
+            trailersFuture.doComplete(trailers);
+        } else {
+            this.trailersFuture.doComplete(trailers);
+        }
+    }
+
+    protected void maybeCompleteHeaders(@Nullable Throwable cause) {
+        if (trailersFuture == null) {
+            trailersFutureUpdater.compareAndSet(this, null, EMPTY_TRAILERS_FUTURE);
+        }
+    }
```

Comment:
How about cleaningup when the `trailersFuture` is not null?
```suggestion
    protected void maybeCompleteHeaders(@Nullable Throwable cause) {
        if (trailersFuture == null) {
            if (trailersFutureUpdater.compareAndSet(this, null, EMPTY_TRAILERS_FUTURE)) {
                return;
            }
        }
        
        final HeadersFuture<HttpHeaders> trailersFuture = this.trailersFuture;
        assert trailersFuture != null;
        trailersFuture.doComplete(HttpHeaders.of());
    }
```